File: Python/Kth Smallest Element in a BST.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    """
        :type root: TreeNode
        :type k: int
        :rtype: int
        """
    node = None
    def kthSmallest(self, root, k):
        self.scan(root, k)
        return Solution.node.val
    
    def scan(self, root, k):
        if root == None:
            return k;
        
        k = self.scan(root.left, k)
        if k == -1:
            return k;
        
        k -= 1
        if k == 0:
            Solution.node = root
            k -= 1
            return k

        return self.scan(root.right, k)

File: Java/Kth Smallest Element in a BST.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    private TreeNode node = null;
    
    public int kthSmallest(TreeNode root, int k) {
        scan(root, k);
        return node.val;
    }
    
    private int scan(TreeNode root, int k) {
        if (root == null) {
            return k;
        }
        k = scan(root.left, k);
        if (k == -1)
            return k;
        
        k -= 1;
        if (k == 0) {
            node = root;
            k -= 1;
            return k;
        }
        
        return scan(root.right, k);
    }
}